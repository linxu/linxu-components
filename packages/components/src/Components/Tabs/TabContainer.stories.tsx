import React from "react";
import { Story } from "@storybook/react";
import { ThemeProvider } from "styled-components";
import { getTheme } from "../../theme/getTheme";
import { ExomeTabs, ExomeTabsProps, Tabs } from "./Tabs";

export default {
  title: "Components/Tabs",
  component: Tabs,
};

export const Minimal: Story<ExomeTabsProps> = (props: ExomeTabsProps) => (
  <ThemeProvider theme={getTheme()}>
    <ExomeTabs {...props} />
  </ThemeProvider>
);

Minimal.args = {
  tabIndex: 2,
  enabledTabIndex: false,
  handleChangeTab: (tab: number) => {
    // eslint-disable-next-line no-console
    console.log(`${tab}`);
  },
  items: [
    {
      label: "Zero",
      width: 100,
    },
    {
      label: "First",
      width: 100,
    },
    {
      label: "Second",
      width: 100,
    },
  ],
} as ExomeTabsProps;
