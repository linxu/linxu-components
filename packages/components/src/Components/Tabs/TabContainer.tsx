import styled from "styled-components";
import { Tab, TabProps } from "@material-ui/core";

export const TabContainer = styled(Tab).attrs((props) => ({
  "data-testid": props["data-testid"],
  classes: {
    root: "tab-container",
    selected: "tab-selected",
    wrapper: "tab-icon-label-container",
  },
}))<TabProps & { width?: string | number }>`
  &.tab-container {
    font-family: ${({ theme }) => theme.typography.main};
    ${({ theme }) => theme.typography.fontSize.medium}
    color: ${({ theme }) => theme.colors.mediumGray};
    font-weight: bold;
    text-transform: capitalize;
    min-width: 88px;
    min-height: 40px;
    padding: 0px;
    width: ${({ width }) => (width ? `${width}px` : "auto")};
  }
  &.tab-selected {
    color: ${({ theme }) => theme.colors.secondary};
    background: ${({ theme }) => theme.colors.highlightPrimary};
    svg {
      fill: ${({ theme }) => theme.colors.secondary};
    }
  }
  .tab-icon-label-container {
    flex-direction: row;
    align-items: center;
  }
  svg {
    margin-right: 5px;
    margin-bottom: 0px !important;
  }
  padding: 0px;
  width: auto;
`;
