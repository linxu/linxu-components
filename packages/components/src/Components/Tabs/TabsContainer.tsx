import React from "react";
import styled, { css } from "styled-components";
import { Tabs, TabsProps } from "@material-ui/core";

interface Props extends TabsProps {
  onChange: any;
  tabWidth?: number | string;
  tabLeft: number;
}

export const TabsContainer = styled(
  ({ tabLeft, tabWidth, ...props }: Props) => <Tabs {...props} />
).attrs(() => ({
  classes: {
    root: "tabs-container",
    indicator: "tabs-indicator",
  },
}))<Props>`
  &.tabs-container {
    min-height: 33px;
    border-bottom: 1px solid ${({ theme }) => theme.colors.lightGray};
  }
  .tabs-indicator {
    height: 2px;
    background: ${({ theme }) => theme.colors.primary};
    ${(props) =>
      props.tabWidth &&
      css`
        width: ${props.tabWidth}px !important;
      `};
    ${(props) =>
      // props.tabLeft &&
      css`
        left: ${props.tabLeft}px !important;
      `};
  }
`;
