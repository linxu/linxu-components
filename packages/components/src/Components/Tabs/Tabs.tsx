import React, { useEffect, useState } from "react";
import { TabsContainer } from "./TabsContainer";
import { TabContainer } from "./TabContainer";

interface TabModel {
  label: string;
  width: number;
  icon?: React.ReactElement;
}

export interface ExomeTabsProps {
  // Callback fired when the tab is selected
  handleChangeTab: (tab: number) => void;
  // Set child tab data,
  // e.g.: [{label: "Data", width: 100, icon: icon}, {label: "Events", width: 100, icon: icon}]
  items?: TabModel[];
  // Selected tab index
  tabIndex?: number;
  enabledTabIndex?: boolean;
}

// Tabs make it easy to explore and switch between different views. <br>
// Other Props Reference https://material-ui.com/api/tabs/
export const Tabs = ({
  handleChangeTab,
  items,
  tabIndex = 0,
  enabledTabIndex = true,
}: ExomeTabsProps) => {
  const [tabWidth, setTabWidth] = useState<string | number | undefined>(
    items ? items[tabIndex].width : "auto"
  );
  const [tabValue, setTabValue] = useState(0);
  const [tabLeft, setTabLeft] = useState(0);
  useEffect(() => {
    if (enabledTabIndex) {
      let offset = 0;
      if (items) {
        items.forEach((item, idx) => {
          if (idx < tabIndex) {
            if (item.width) {
              offset += item.width;
            }
          }
        });
        setTabLeft(offset);
      }
    }
  }, [tabIndex]);

  useEffect(() => {
    if (!enabledTabIndex) {
      let offset = 0;
      if (items) {
        items.forEach((item, idx) => {
          if (idx < tabValue) {
            if (item.width) {
              offset += item.width;
            }
          }
        });
        setTabLeft(offset);
      }
    }
  }, [tabValue]);

  return (
    <TabsContainer
      value={enabledTabIndex ? tabIndex : tabValue}
      tabLeft={tabLeft}
      tabWidth={tabWidth}
      onChange={(event, value) => {
        setTabWidth(items!![value].width);
        setTabValue(value);
        handleChangeTab(value);
      }}
    >
      {(items || []).map((item, index) => (
        <TabContainer
          data-testid={`tabs-item-${index}`}
          key={item.label}
          disableRipple
          label={item.label}
          width={item.width}
          icon={item.icon}
        />
      ))}
    </TabsContainer>
  );
};

export const ExomeTabs = Tabs;
