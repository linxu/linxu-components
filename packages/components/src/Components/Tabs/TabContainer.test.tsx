import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";

import { Minimal } from "./TabContainer.stories";
import { ExomeTabsProps } from "./Tabs";

it("change tab", async () => {
  const tree = render(
    <Minimal {...(Minimal.args as ExomeTabsProps)} />
  );
  const item = await screen.findByTestId("tabs-item-0");
  await fireEvent.click(item);
  expect(tree).toMatchSnapshot();
});

it("without items", async () => {
  const tree = render(
    <Minimal
      handleChangeTab={jest.fn()}
      enabledTabIndex={false}
      items={undefined}
    />
  );
  expect(tree).toMatchSnapshot();
});
