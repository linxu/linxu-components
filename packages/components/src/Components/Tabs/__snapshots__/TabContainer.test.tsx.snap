// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change tab 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTabs-root tabs-container sc-bdfBQB cEpzDE"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed"
          style="overflow: hidden;"
        >
          <div
            class="MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW Mui-selected tab-selected"
              data-testid="tabs-item-0"
              role="tab"
              tabindex="0"
              type="button"
              width="100"
            >
              <span
                class="MuiTab-wrapper tab-icon-label-container"
              >
                Zero
              </span>
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW"
              data-testid="tabs-item-1"
              role="tab"
              tabindex="-1"
              type="button"
              width="100"
            >
              <span
                class="MuiTab-wrapper tab-icon-label-container"
              >
                First
              </span>
            </button>
            <button
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW"
              data-testid="tabs-item-2"
              role="tab"
              tabindex="-1"
              type="button"
              width="100"
            >
              <span
                class="MuiTab-wrapper tab-icon-label-container"
              >
                Second
              </span>
            </button>
          </div>
          <span
            class="PrivateTabIndicator-root-1 PrivateTabIndicator-colorSecondary-3 MuiTabs-indicator tabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTabs-root tabs-container sc-bdfBQB cEpzDE"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW Mui-selected tab-selected"
            data-testid="tabs-item-0"
            role="tab"
            tabindex="0"
            type="button"
            width="100"
          >
            <span
              class="MuiTab-wrapper tab-icon-label-container"
            >
              Zero
            </span>
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW"
            data-testid="tabs-item-1"
            role="tab"
            tabindex="-1"
            type="button"
            width="100"
          >
            <span
              class="MuiTab-wrapper tab-icon-label-container"
            >
              First
            </span>
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root tab-container MuiTab-textColorInherit sc-gsTEea NpXkW"
            data-testid="tabs-item-2"
            role="tab"
            tabindex="-1"
            type="button"
            width="100"
          >
            <span
              class="MuiTab-wrapper tab-icon-label-container"
            >
              Second
            </span>
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root-1 PrivateTabIndicator-colorSecondary-3 MuiTabs-indicator tabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`without items 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTabs-root tabs-container sc-bdfBQB fOgnxW"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed"
          style="overflow: hidden;"
        >
          <div
            class="MuiTabs-flexContainer"
            role="tablist"
          />
          <span
            class="PrivateTabIndicator-root-5 PrivateTabIndicator-colorSecondary-7 MuiTabs-indicator tabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTabs-root tabs-container sc-bdfBQB fOgnxW"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        />
        <span
          class="PrivateTabIndicator-root-5 PrivateTabIndicator-colorSecondary-7 MuiTabs-indicator tabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
